File: scripts/export_schemas.py
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

from dere_shared.config import get_config_schema
from dere_shared.llm_schemas import LLM_SCHEMA_REGISTRY


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[1]


def _to_snake(name: str) -> str:
    step1 = re.sub(r"(.)([A-Z][a-z]+)", r"\1_\2", name)
    step2 = re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", step1)
    return step2.lower()


def _write_json(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(
        json.dumps(payload, indent=2, sort_keys=True, ensure_ascii=True) + "\n",
        encoding="utf-8",
    )


def export_llm_schemas(base_dir: Path) -> list[Path]:
    output_paths: list[Path] = []
    for name, model in LLM_SCHEMA_REGISTRY.items():
        schema = model.model_json_schema()
        filename = f"{_to_snake(name)}.schema.json"
        path = base_dir / "llm" / filename
        _write_json(path, schema)
        output_paths.append(path)
    return output_paths


def export_config_schema(base_dir: Path) -> Path:
    schema = get_config_schema()
    path = base_dir / "config" / "dere_config.schema.json"
    _write_json(path, schema)
    return path


def main() -> None:
    base_dir = _repo_root() / "schemas"
    llm_outputs = export_llm_schemas(base_dir)
    config_output = export_config_schema(base_dir)

    summary = [
        "Exported schemas:",
        *(f"- {path.relative_to(_repo_root())}" for path in llm_outputs),
        f"- {config_output.relative_to(_repo_root())}",
    ]
    print("\n".join(summary))


if __name__ == "__main__":
    main()

File: scripts/export_schemas.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { z } from "zod";

import {
  AmbientEngagementDecisionSchema,
  AmbientMissionDecisionSchema,
  AppraisalOutputSchema,
  ExplorationOutputSchema,
  ScheduleParseResultSchema,
  SessionTitleResultSchema,
} from "../packages/shared-llm/src/schemas.ts";

type JsonRecord = Record<string, unknown>;

const scriptDir = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(scriptDir, "..");

async function writeJson(path: string, payload: JsonRecord): Promise<void> {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, `${JSON.stringify(payload, null, 2)}\n`, "utf-8");
}

const registry = [
  {
    name: "appraisal_output",
    title: "AppraisalOutput",
    schema: AppraisalOutputSchema,
  },
  {
    name: "ambient_engagement_decision",
    title: "AmbientEngagementDecision",
    schema: AmbientEngagementDecisionSchema,
  },
  {
    name: "ambient_mission_decision",
    title: "AmbientMissionDecision",
    schema: AmbientMissionDecisionSchema,
  },
  {
    name: "exploration_output",
    title: "ExplorationOutput",
    schema: ExplorationOutputSchema,
  },
  {
    name: "schedule_parse_result",
    title: "ScheduleParseResult",
    schema: ScheduleParseResultSchema,
  },
  {
    name: "session_title_result",
    title: "SessionTitleResult",
    schema: SessionTitleResultSchema,
  },
];

async function exportLlmSchemas(baseDir: string): Promise<string[]> {
  const outputs: string[] = [];
  for (const entry of registry) {
    const jsonSchema = z.toJSONSchema(entry.schema) as JsonRecord;
    if (entry.title) {
      jsonSchema.title = entry.title;
    }
    const path = join(baseDir, "llm", `${entry.name}.schema.json`);
    await writeJson(path, jsonSchema);
    outputs.push(path);
  }
  return outputs;
}

async function exportConfigSchema(baseDir: string): Promise<string> {
  const path = join(baseDir, "config", "dere_config.schema.json");
  const raw = await readFile(path, "utf-8");
  const schema = JSON.parse(raw) as JsonRecord;
  await writeJson(path, schema);
  return path;
}

async function main(): Promise<void> {
  const baseDir = join(repoRoot, "schemas");
  const llmOutputs = await exportLlmSchemas(baseDir);
  const configOutput = await exportConfigSchema(baseDir);

  const summary = [
    "Exported schemas:",
    ...llmOutputs.map((path) => `- ${path.slice(repoRoot.length + 1)}`),
    `- ${configOutput.slice(repoRoot.length + 1)}`,
  ];
  console.log(summary.join("\n"));
}

await main();
